harness/determined/_core/_distributed.py

```diff
@@ -215,11 +219,13 @@ def _zmq_gather(self, stuff: Any) -> Optional[List]:
             return [stuff]
         logging.debug(f"Worker {self.get_rank()} beginning zmq gather.")
         if self._is_chief:
-            worker_stuff, _ = self._chief_zmq.gather_with_polling(lambda: None)
+            worker_stuff_ranked, _ = self._chief_zmq.gather_with_polling(lambda: None)
+            worker_stuff_ranked.sort(key=_tuple_key_function)
+            worker_stuff = [value for value, _ in worker_stuff_ranked]
             self._chief_zmq.broadcast(None)
             out = [stuff, *worker_stuff]  # type: Optional[List]
         else:
-            self._worker_zmq.send(stuff)
+            self._worker_zmq.send((stuff, self.get_rank()))
```

Comment:
You could reorder this tuple so you don't have to provide a key function, I believe.

I'm pretty sure that the ordering of tuples by default is based on the default sorting of their first elements.